atom/browser/io_thread.h

```diff
@@ -53,6 +62,13 @@ class IOThread : public content::BrowserThreadDelegate {
   network::mojom::HttpAuthStaticParamsPtr http_auth_static_params_;
   network::mojom::HttpAuthDynamicParamsPtr http_auth_dynamic_params_;
 
+  // |lock_| protects access to |request_context_getters_|.
+  base::Lock lock_;
+
+  // List of all request contexts that needs to be notified when
+  // IO thread is shutting down.
+  std::set<atom::URLRequestContextGetter*> request_context_getters_;
```

Comment:
Hm... it seems like there might be a simpler way to handle the lifetimes of these objects. How does Chrome do it?

Reply:
We can't align with Chrome's behavior wrt these objects, because the `URLRequestContextGetter` is bound to `BrowserContext` , but we manage `BrowserContext` lifetime using the session module exposed to the users and these objects unless explicitly destroyed, they will get destroyed on app exit using the destructors registered in [AtomBrowserMainParts::PostMainMessageLoopRun](https://github.com/electron/electron/blob/master/atom/browser/atom_browser_main_parts.cc#L463-L473)

Previously the `NetworkService` object didn't expect the primary network context to be destroyed, since there is only one per app lifetime. But now we need to make sure its destroyed, and destroying a primary network context requires all other network context objects destroyed before it. Although the IO thread cleanup happens after `PostMainMessageLoopRun` https://cs.chromium.org/chromium/src/content/browser/browser_main_loop.cc?l=1113 , sometimes the task we post to destroy `URLRequestContext` in `atom::URLRequestContextGetter::Handle::ShutdownOnUIThread` might not happen around the time `IOThread::Cleanup` is called, so the destruction order is messed up.

This change makes sure to track and delete those objects, its only a problem we face because we have exposed a dependency of these objects to user land. Also this is only a problem for pre network service code path, this will be going away soon. I think its fine to maintain it for the short time :)